### ISSUE TITLE
Can EF Core 7 translate Concat (UNION ALL) with a value?

### ISSUE THREAD

**Issue by DaveCousineau** [dotnet/efcore#32337]
This is specifically for EF Core 7 (.NET 6). As far as I can tell `Concat` is supposed to be supported and it is supposed to translate to `UNION ALL`.

In EF 6 I used to be able to use `Concat` (so the `Queryable.Concat` extension method) something like this:

```c#
var results =
   context
   .Entities
   .Select(e => e.ID)
   .Concat(new[] { 1234 })
   .Distinct()
   .OrderByDescending(x => x)
   .ToArray();
```

But it says that it cannot translate this query. (I know I can switch to `IEnumerable` to do that part locally, which I will do if I have to, but it's slightly better if it's done DB-side and then I could also still compose against the `IQueryable` further if I needed to.)

```
System.InvalidOperationException: The LINQ expression 'DbSet<Entity>()
    .Select(e => e.ID)
    .Concat(__p_0)' could not be translated. Either rewrite the query in a form that can be translated, or switch to client evaluation explicitly by inserting a call to 'AsEnumerable', 'AsAsyncEnumerable', 'ToList', or 'ToListAsync'. See https://go.microsoft.com/fwlink/?linkid=2101038 for more information.
    at Microsoft.EntityFrameworkCore.Query.Internal.NavigationExpandingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
   at System.Linq.Expressions.MethodCallExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
   at Microsoft.EntityFrameworkCore.Query.Internal.NavigationExpandingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
   at System.Linq.Expressions.MethodCallExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
   at Microsoft.EntityFrameworkCore.Query.Internal.NavigationExpandingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
   at System.Linq.Expressions.MethodCallExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
   at Microsoft.EntityFrameworkCore.Query.Internal.NavigationExpandingExpressionVisitor.Expand(Expression query)
   at Microsoft.EntityFrameworkCore.Query.QueryTranslationPreprocessor.Process(Expression query)
   at Microsoft.EntityFrameworkCore.Query.QueryCompilationContext.CreateQueryExecutor[TResult](Expression query)
   at Microsoft.EntityFrameworkCore.Storage.Database.CompileQuery[TResult](Expression query, Boolean async)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.CompileQueryCore[TResult](IDatabase database, Expression query, IModel model, Boolean async)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.<>c__DisplayClass9_0`1.<Execute>b__0()
   at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQuery[TResult](Object cacheKey, Func`1 compiler)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.Execute[TResult](Expression query)
   at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryProvider.Execute[TResult](Expression expression)
   at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryable`1.GetEnumerator()
   at System.Collections.Generic.LargeArrayBuilder`1.AddRange(IEnumerable`1 items)
   at System.Collections.Generic.EnumerableHelpers.ToArray[T](IEnumerable`1 source)
   at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)
```

I suppose maybe `Concat` is only supported with two `IQueryable`s or something, although the parameter is an `IEnumerable` not an `IQueryable`.

I seem to be able to write something like the following, so it seems strange that you would not support a value enumerable.

```c#
// this sort of works except I guess it would not return 1234 if the set was empty
var results =
   context
   .Entities
   .Select(e => e.ID)
   .Concat(
      context
      .Entities
      .Select(e => 1234)
      .Take(1)
   ).ToArray();
```

Thanks

EF Core version: 7.0.14
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework: .NET 6
Operating system: Windows 10 Pro 22H2
IDE: JetBrains Rider 2023.2.3

**Comment by roji** [comment]
This query now works in the recently-released 8.0, thanks to the work we did on primitive collections. Please give that a try.

BTW while testing this I noticed that the inline collection gets translated as a parameter instead of inline (with VALUES), opened #32339 to track that.

**Comment by DaveCousineau** [comment]
Ok, thanks for the reply. I don't think I will be upgrading to .NET 7/EF Core 8 right now but I will tag this for later when I do.

Thanks
